solr/core/src/java/org/apache/solr/search/KeepAliveRegenerator.java

```diff
@@ -0,0 +1,593 @@
+    }
+
+    KeepAliveSegAwareValue metaEntry = (KeepAliveSegAwareValue) oldVal;
+    final long extantTimestamp = metaEntry.accessTimestampNanos;
+    long lastAccessAgo = System.nanoTime() - extantTimestamp;
+    if (lastAccessAgo > regenKeepAliveNanos) {
+      // it has been long enough since this was last accessed that we don't want to carry it forward
+      return true;
+    }
+
+    final Query query = (Query) oldKey;
+
+    @SuppressWarnings("unchecked")
+    SolrCache<Query, KeepAliveSegAwareValue> c =
+        (SolrCache<Query, KeepAliveSegAwareValue>) newCache;
+
+    if (metaEntry.ref.get().getKey().registerOverlap(newSearcher.getSegmentMap()) < overlapThreshold
+        || isCrossDoc(query)) {
+      if (lastAccessAgo < eagerKeepAliveNanos) {
+        // if we meet the criterion for eager warming, do it here (not leveraging segment-aware)
+        c.computeIfAbsent(
+            query,
+            (q) -> {
+              SegmentMap segMap = newSearcher.getSegmentMap();
+              DocSet docSet = newSearcher.getDocSetNC(query, null);
+              return new KeepAliveSegAwareValue(
+                  segMap, docSet, extantTimestamp, partialHits, partialHitsRatio);
+            });
+      }
+      return true;
+    }
+
+    if (!preferLazy && lastAccessAgo < eagerKeepAliveNanos) {
+      // if we're doing eager warming, do it here even for segment-aware entries. We can
+      // still benefit from the old cache entry.
+      c.computeIfAbsent(
+          query,
+          (q) -> {
+            AbstractMap.SimpleImmutableEntry<SegmentMap, CompletableFuture<DocSet>> ref =
+                metaEntry.ref.get();
+            Query frankenstein;
+            try {
+              frankenstein =
+                  new FrankensteinQuery(query, ref.getKey().segments, ref.getValue().get());
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+              throw new RuntimeException(e);
+            } catch (ExecutionException e) {
+              Throwable cause = e.getCause();
+              if (cause instanceof IOException) {
+                throw (IOException) cause;
+              } else {
+                throw new RuntimeException(e);
+              }
+            }
+            DocSet docSet = newSearcher.getDocSetNC(frankenstein, null);
+            return new KeepAliveSegAwareValue(
+                newSearcher.getSegmentMap(),
+                docSet,
+                extantTimestamp,
+                partialHits,
+                partialHitsRatio);
+          });
+      return true;
+    }
+
+    // the lazy way. Just pass the stale cache entry along in case it's queried later
+    c.computeIfAbsent(
+        query, (q) -> new KeepAliveSegAwareValue(metaEntry, partialHits, partialHitsRatio));
+    return true;
+  }
+
+  private static class FrankensteinQuery extends Query {
+
+    private final Query backing;
+    private final Map<IndexReader.CacheKey, SegmentMap.Segment> segs;
```

Comment:
Minor: can we rename this to previousSegments or staleSegments? (i like previousSegments better since most of these segments aren't really stale)

I was confused for a bit when I saw this class was instantiated with the previous segment map at first 😅